Workiva/over_react
# CPLAT-12196: @Component2 annotation parity for function components

## Motivation
  
The `@Component2` annotation can't be used with function components, preventing consumers from specifying arguments like `isWrapper` and `subtypeOf`.

## Changes
- Add `setTypeMeta` extension method to `UiFactory`, providing parity with the `@Component2` annotation for function components
    - Add tests
- Tweak arguments to private `setComponentTypeMeta` and `ComponentTypeMeta` to use raw component types in case the `ReactComponentFactoryProxy` isn't available
    - Increase test coverage for these utilities

Examples of equivalent configuration for class-based vs function components:
```dart
// Class components:

@Component2(isWrapper: true)
class FooComponent extends UiComponent2<FooProps> { ... }

@Component2(subtypeOf: FooComponent)
class BarComponent ... {...}

// Function components:

UiFactory<FooProps> Foo = uiFunction(...)
  ..setComponentMeta(isWrapper: true);

UiFactory<FooProps> Bar = uiFunction(...)
  ..setComponentMeta(subtypeOfFactory: Foo);
```

#### Release Notes
- Add extension method `UiFactory.setComponentMeta`, which brings `@Component2` annotation configuration to function components

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        - Ensure test branch WSD is passing: https://github.com/Workiva/web_skin_dart/pull/1537
            - WSD has extensive usage of component subtyping, and is a good place to test for regressions related to these changes

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
